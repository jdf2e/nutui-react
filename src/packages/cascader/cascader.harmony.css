.nut-cascader {
  width: 100%;
  font-size: 14px;
}
.nut-cascader .nut-tabs-titles {
  padding: 0 10px;
  background: #ffffff;
}
.nut-cascader .nut-tabs-titles-item {
  flex: initial;
  min-width: auto;
  width: auto;
  padding: 0 10px;
  white-space: nowrap;
}
.nut-cascader .nut-tabpane {
  padding: 0;
  background: #ffffff;
}
.nut-cascader-pane {
  display: block;
  width: 100%;
  padding-top: 10px;
  height: 342px;
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
}
.nut-cascader-item {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 10px 20px;
  margin: 0px;
  border-bottom: 0px solid rgba(0, 0, 0, 0.06);
  font-size: 14px;
  color: #1a1a1a;
  cursor: pointer;
}
.nut-cascader-item.disabled {
  opacity: 0.6;
  cursor: not-allowed;
}
.nut-cascader-item.active:not(.disabled) {
  color: #fa2c19;
}
.nut-cascader-item.active .nut-cascader-item-icon-check {
  visibility: visible;
  color: inherit;
}
.nut-cascader-item-title {
  flex: 1;
}
.nut-cascader .nut-icon-checklist {
  margin-left: 10px;
  visibility: hidden;
}

[dir=rtl] .nut-cascader .nut-icon-checklist {
  margin-left: 0;
  margin-right: 10px;
}

.nut-rtl .nut-cascader .nut-icon-checklist {
  margin-left: 0;
  margin-right: 10px;
}