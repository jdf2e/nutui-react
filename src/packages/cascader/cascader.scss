@import '../popup/popup.scss';
@import '../tabs/tabs.scss';

.nut-cascader {
  width: 100%;
  font-size: $cascader-font-size;

  .nut-tabs-titles {
    padding: $cascader-tabs-item-padding;
    background: $color-background-overlay;
  }

  .nut-tabs-titles-item {
    flex: initial;
    min-width: auto;
    width: auto;
    padding: $cascader-tabs-item-padding;
    white-space: nowrap;
  }

  .nut-tabpane {
    padding: 0;
    background: $color-background-overlay;
  }

  &-pane {
    display: block;
    width: 100%;
    padding-top: $cascader-pane-paddingTop;
    height: $cascader-pane-height;
    overflow-y: auto;
    -webkit-overflow-scrolling: touch;
  }

  &-item {
    display: flex;
    align-items: center;
    justify-content: center;
    // height: $cascader-item-height;
    padding: $cascader-item-padding;
    margin: $cascader-item-margin;
    border-bottom: $cascader-item-border-bottom;
    font-size: $cascader-item-font-size;
    color: $cascader-item-color;
    cursor: pointer;

    &.disabled {
      opacity: 0.6;
      cursor: not-allowed;
    }

    &.active {
      &:not(.disabled) {
        color: $cascader-item-active-color;
      }

      .nut-cascader-item-icon-check {
        visibility: visible;
        color: inherit;
      }
    }
  }

  &-item-title {
    flex: 1;
  }

  .nut-icon-checklist {
    margin-left: $cascader-icon-checklist-marginLeft;
    visibility: hidden;
  }
}

[dir='rtl'] .nut-cascader {
  .nut-icon-checklist {
    margin-left: 0;
    margin-right: $cascader-icon-checklist-marginLeft;
  }
}
