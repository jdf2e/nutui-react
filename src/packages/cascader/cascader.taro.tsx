import React, {
  ForwardRefRenderFunction,
  PropsWithChildren,
  isValidElement,
  useState,
  useEffect,
  ReactNode,
  useImperativeHandle,
} from 'react'
import classNames from 'classnames'
import { Loading, Checklist } from '@nutui/icons-react-taro'
import { ScrollView } from '@tarojs/components'
import { Popup, PopupProps } from '@/packages/popup/popup.taro'
import { Tabs } from '@/packages/tabs/tabs.taro'
import { convertListToOptions } from './helper'
import {
  CascaderPane,
  CascaderOption,
  CascaderValue,
  CascaderOptionKey,
  CascaderFormat,
} from './types'
import Tree from './tree'
import { ComponentDefaults } from '@/utils/typings'
import { usePropsValue } from '@/utils/use-props-value'
import { useConfig } from '@/packages/configprovider/configprovider.taro'

export interface CascaderProps
  extends Pick<
    PopupProps,
    | 'className'
    | 'style'
    | 'closeIcon'
    | 'closeable'
    | 'title'
    | 'left'
    | 'closeIconPosition'
    | 'onClose'
  > {
  popup: boolean
  popupProps: Partial<
    Omit<
      PopupProps,
      | 'closeIcon'
      | 'closeable'
      | 'title'
      | 'left'
      | 'closeIconPosition'
      | 'onClose'
    >
  >
  visible: boolean // popup visible
  activeColor: string
  activeIcon: string
  options: CascaderOption[]
  value?: CascaderValue
  defaultValue?: CascaderValue
  optionKey: CascaderOptionKey
  format: Record<string, string | number | null>
  closeable: boolean
  closeIconPosition: string
  closeIcon: ReactNode
  lazy: boolean
  onLoad: (node: any, resolve: any) => void
  onChange: (value: CascaderValue, params?: any) => void
  onPathChange: (value: CascaderValue, params: any) => void
}

export type CascaderActions = {
  open: () => void
  close: () => void
}

const defaultProps = {
  ...ComponentDefaults,
  activeColor: '',
  activeIcon: 'checklist',
  popup: true,
  options: [],
  optionKey: { textKey: 'text', valueKey: 'value', childrenKey: 'children' },
  format: {},
  closeable: false,
  closeIconPosition: 'top-right',
  closeIcon: 'close',
  lazy: false,
  onLoad: () => {},
  onClose: () => {},
  onChange: () => {},
  onPathChange: () => {},
} as unknown as CascaderProps
const InternalCascader: ForwardRefRenderFunction<
  unknown,
  PropsWithChildren<Partial<CascaderProps>>
> = (props, ref) => {
  const { locale } = useConfig()
  const {
    className,
    style,
    activeColor,
    activeIcon,
    popup,
    popupProps = {},
    visible,
    options,
    value,
    defaultValue,
    optionKey,
    format,
    closeable,
    closeIconPosition,
    closeIcon,
    lazy,
    title,
    left,
    onLoad,
    onClose,
    onChange,
    onPathChange,
  } = { ...defaultProps, ...props }

  const [tabvalue, setTabvalue] = useState('c1')
  const [optionsData, setOptionsData] = useState<CascaderPane[]>([])
  const isLazy = () => state.configs.lazy && Boolean(state.configs.onLoad)

  const [innerValue, setInnerValue] = usePropsValue<CascaderValue>({
    value,
    defaultValue,
    finalValue: defaultValue,
  })
  const [innerVisible, setInnerVisible] = usePropsValue<boolean>({
    value: visible,
    defaultValue: undefined,
    finalValue: false,
  })
  const actions: CascaderActions = {
    open: () => {
      setInnerVisible(true)
    },
    close: () => {
      setInnerVisible(false)
    },
  }
  useImperativeHandle(ref, () => actions)

  const [state] = useState({
    optionsData: [] as any,
    panes: [
      {
        nodes: [] as any,
        selectedNode: [] as CascaderOption | null,
        paneKey: '',
      },
    ],
    tree: new Tree([], {}),
    tabsCursor: 0, // 选中的tab项
    initLoading: false,
    currentProcessNode: [] as CascaderOption | null,
    configs: {
      lazy,
      onLoad,
      optionKey,
      format,
    },
    lazyLoadMap: new Map(),
  })

  const classPrefix = classNames(`nut-cascader`)
  const classesPane = classNames({
    [`${classPrefix}-pane`]: true,
  })

  useEffect(() => {
    initData()
  }, [options, format])

  useEffect(() => {
    syncValue()
  }, [value])

  const initData = async () => {
    // 初始化开始处理数据
    state.lazyLoadMap.clear()
    if (format && Object.keys(format).length > 0) {
      state.optionsData = convertListToOptions(
        options as CascaderOption[],
        format as CascaderFormat
      )
    } else {
      state.optionsData = options
    }
    state.tree = new Tree(state.optionsData as CascaderOption[], {
      value: state.configs.optionKey.valueKey,
      text: state.configs.optionKey.textKey,
      children: state.configs.optionKey.childrenKey,
    })
    if (isLazy() && !state.tree.nodes.length) {
      await invokeLazyLoad({
        root: true,
        loading: true,
        text: '',
        value: '',
      })
    }
    state.panes = [
      {
        nodes: state.tree.nodes,
        selectedNode: null,
        paneKey: 'c1',
      },
    ]
    syncValue()
    setOptionsData(state.panes)
  }
  // 处理有默认值时的数据
  const syncValue = async () => {
    const currentValue = innerValue

    if (
      currentValue === undefined ||
      ![defaultValue, value].includes(currentValue) ||
      !state.tree.nodes.length
    ) {
      return
    }

    if (currentValue.length === 0) {
      state.tabsCursor = 0
      return
    }

    let needToSync = currentValue

    if (isLazy() && Array.isArray(currentValue) && currentValue.length) {
      needToSync = []
      const parent: any = state.tree.nodes.find(
        (node) => node.value === currentValue[0]
      )

      if (parent) {
        needToSync = [parent.value]
        state.initLoading = true

        const last = await currentValue
          .slice(1)
          .reduce(async (p: Promise<CascaderOption | void>, value) => {
            const parent = await p
            await invokeLazyLoad(parent)
            const node: any = parent?.children?.find(
              (item: any) => item.value === value
            )
            if (node) {
              needToSync.push(value)
            }
            return Promise.resolve(node)
          }, Promise.resolve(parent))
        await invokeLazyLoad(last)
        state.initLoading = false
      }
    }

    if (needToSync.length && [defaultValue, value].includes(currentValue)) {
      const pathNodes = state.tree.getPathNodesByValue(needToSync)
      pathNodes.forEach((node, index) => {
        state.tabsCursor = index
        // 当有默认值时，不触发 chooseItem 里的 emit 事件
        chooseItem(node, true)
      })
    }
  }

  const invokeLazyLoad = async (node?: CascaderOption | void) => {
    if (!node) {
      return
    }

    if (!state.configs.onLoad) {
      node.leaf = true
      return
    }

    if (
      state.tree.isLeaf(node, isLazy()) ||
      state.tree.hasChildren(node, isLazy())
    ) {
      return
    }

    node.loading = true

    const parent = node.root ? null : node
    let lazyLoadPromise = state.lazyLoadMap.get(node)

    if (!lazyLoadPromise) {
      lazyLoadPromise = new Promise((resolve) => {
        // 外部必须resolve
        state.configs.onLoad?.(node, resolve)
      })
      state.lazyLoadMap.set(node, lazyLoadPromise)
    }

    const nodes: CascaderOption[] | void = await lazyLoadPromise

    if (Array.isArray(nodes) && nodes.length > 0) {
      state.tree.updateChildren(nodes, parent)
    } else {
      // 如果加载完成后没有提供子节点，作为叶子节点处理
      node.leaf = true
    }
    node.loading = false
    state.lazyLoadMap.delete(node)
  }

  const close = () => {
    setInnerVisible(false)
    onClose && onClose()
  }

  const closePopup = () => {
    close()
  }

  /* type: 是否是静默模式，是的话不触发事件
  tabsCursor: tab的索引 */
  const chooseItem = async (node: CascaderOption, type: boolean) => {
    if ((!type && node.disabled) || !state.panes[state.tabsCursor]) {
      return
    }
    // 如果没有子节点
    if (state.tree.isLeaf(node, isLazy())) {
      node.leaf = true
      state.panes[state.tabsCursor].selectedNode = node
      state.panes = state.panes.slice(0, (node.level as number) + 1)
      if (!type) {
        const pathNodes = state.panes.map((item) => item.selectedNode)
        const optionParams = pathNodes.map((item: any) => item.value)
        onChange(optionParams, pathNodes)
        onPathChange?.(optionParams, pathNodes)
        setInnerValue(optionParams)
      }
      setOptionsData(state.panes)
      close()
      return
    }
    // 如果有子节点，滑到下一个
    if (state.tree.hasChildren(node, isLazy())) {
      const level = (node.level as number) + 1

      state.panes[state.tabsCursor].selectedNode = node
      state.panes = state.panes.slice(0, level)
      state.tabsCursor = level
      state.panes.push({
        nodes: node.children || [],
        selectedNode: null,
        paneKey: `c${state.tabsCursor + 1}`,
      })
      setOptionsData(state.panes)
      setTabvalue(`c${state.tabsCursor + 1}`)

      if (!type) {
        const pathNodes = state.panes.map((item) => item.selectedNode)
        const optionParams = pathNodes.map((item: any) => item?.value)
        onPathChange?.(optionParams, pathNodes)
      }
      return
    }
    state.currentProcessNode = node
    if (node.loading) {
      return
    }

    await invokeLazyLoad(node)
    if (state.currentProcessNode === node) {
      state.panes[state.tabsCursor].selectedNode = node
      chooseItem(node, type)
    }
    setOptionsData(state.panes)
  }

  const renderItem = (pane: any, node: any, index: number) => {
    const classPrefix2 = 'nut-cascader-item'
    const checked = pane.selectedNode?.value === node.value

    const classes = classNames(
      {
        active: checked,
        disabled: node.disabled,
      },
      classPrefix2
    )

    const classesTitle = classNames({
      [`${classPrefix2}-title`]: true,
    })

    const renderIcon = () => {
      if (checked) {
        if (isValidElement(activeIcon)) {
          return activeIcon
        }
        return (
          <Checklist
            className={`${checked ? `${classPrefix}-icon-check` : ''}`}
          />
        )
      }
      return null
    }

    return (
      <div
        style={{ color: checked ? activeColor : '' }}
        className={classes}
        key={index}
        onClick={() => {
          chooseItem(node, false)
        }}
      >
        <div className={classesTitle}>{node.text}</div>
        {node.loading ? (
          <Loading color="#969799" className="nut-cascader-item-icon-loading" />
        ) : (
          renderIcon()
        )}
      </div>
    )
  }

  const renderTabs = () => {
    return (
      <div className={`${classPrefix} ${className}`} style={style}>
        <Tabs
          value={tabvalue}
          title={() => {
            return optionsData.map((pane, index) => (
              <div
                onClick={() => {
                  setTabvalue(pane.paneKey)
                  state.tabsCursor = index
                }}
                className={`nut-tabs-titles-item ${
                  tabvalue === pane.paneKey ? 'nut-tabs-titles-item-active' : ''
                }`}
                key={pane.paneKey}
              >
                <span className="nut-tabs-titles-item-text">
                  {!state.initLoading &&
                    state.panes.length &&
                    pane?.selectedNode?.text}
                  {!state.initLoading &&
                    state.panes.length &&
                    !pane?.selectedNode?.text &&
                    `${locale.select}`}
                  {!(!state.initLoading && state.panes.length) && 'Loading...'}
                </span>
                <span className="nut-tabs-titles-item-line" />
              </div>
            ))
          }}
        >
          {!state.initLoading && state.panes.length ? (
            optionsData.map((pane) => (
              <Tabs.TabPane key={pane.paneKey} value={pane.paneKey}>
                <ScrollView className={classesPane} scrollY>
                  {pane.nodes?.map((node: any, index: number) =>
                    renderItem(pane, node, index)
                  )}
                </ScrollView>
              </Tabs.TabPane>
            ))
          ) : (
            <Tabs.TabPane>
              <div className={classesPane} />
            </Tabs.TabPane>
          )}
        </Tabs>
      </div>
    )
  }

  return (
    <>
      {popup ? (
        <Popup
          {...popupProps}
          visible={innerVisible}
          position="bottom"
          style={{ overflowY: 'hidden' }}
          round
          closeIcon={closeIcon}
          closeable={closeable}
          closeIconPosition={closeIconPosition}
          title={popup && (title as ReactNode)}
          left={left}
          // todo 只关闭，不处理逻辑。和popup的逻辑不一致。关闭时需要增加是否要处理回调
          onOverlayClick={closePopup}
          onCloseIconClick={closePopup}
        >
          {renderTabs()}
        </Popup>
      ) : (
        renderTabs()
      )}
    </>
  )
}

export const Cascader = React.forwardRef(InternalCascader)

Cascader.displayName = 'NutCascader'
