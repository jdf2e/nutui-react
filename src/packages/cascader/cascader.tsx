import React, {
  ForwardRefRenderFunction,
  PropsWithChildren,
  useState,
  useEffect,
} from 'react'
import classNames from 'classnames'
import Popup from '@/packages/popup'
import { Tabs } from '@/packages/tabs/tabs'
import { CascaderItem } from './cascaderItem'
import { convertListToOptions } from './helper'
import {
  CascaderPane,
  CascaderOption,
  CascaderValue,
  optionKey,
  format,
} from './types'
import Tree from './tree'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'

export interface CascaderProps extends BasicComponent {
  popup: boolean
  visible: boolean // popup 显示状态
  activeColor: string
  activeIcon: string
  options: CascaderOption[]
  value: string[]
  title: string
  optionKey: optionKey
  format: Record<string, string | number | null>
  closeable: boolean
  closeIconPosition: string
  closeIcon: string
  lazy: boolean
  onLoad: (node: any, resolve: any) => void
  onClose?: () => void
  onChange: (value: any, params: any) => void
  onPathChange: (value: any, params: any) => void
}

const defaultProps = {
  ...ComponentDefaults,
  activeColor: '',
  activeIcon: 'checklist',
  popup: true,
  visible: false,
  options: [],
  value: [],
  title: '',
  optionKey: { textKey: 'text', valueKey: 'value', childrenKey: 'children' },
  format: {},
  closeable: false,
  closeIconPosition: 'top-right',
  closeIcon: 'close',
  lazy: false,
  onLoad: () => {},
  onClose: () => {},
  onChange: () => {},
  onPathChange: () => {},
  ...Popup.defaultProps,
} as CascaderProps
const InternalCascader: ForwardRefRenderFunction<
  unknown,
  PropsWithChildren<Partial<CascaderProps>>
> = (props, ref) => {
  const {
    className,
    style,
    popup,
    visible,
    options,
    value,
    title,
    optionKey,
    format,
    closeable,
    closeIconPosition,
    closeIcon,
    lazy,
    onLoad,
    onClose,
    onChange,
    onPathChange,
  } = { ...defaultProps, ...props }

  const [tabvalue, setTabvalue] = useState('c1')
  const [optionsData, setOptionsData] = useState<CascaderPane[]>([])
  const isLazy = () => state.configs.lazy && Boolean(state.configs.onLoad)

  const [state] = useState({
    optionsData: [] as any,
    panes: [
      {
        nodes: [] as any,
        selectedNode: [] as CascaderOption | null,
        paneKey: '',
      },
    ],
    innerValue: value as CascaderValue,
    tree: new Tree([], {}),
    tabsCursor: 0, // 选中的tab项
    initLoading: false,
    currentProcessNode: [] as CascaderOption | null,
    configs: {
      lazy,
      onLoad,
      optionKey,
      format,
    },
    lazyLoadMap: new Map(),
  })

  const classPrefix = classNames(`nut-cascader`)
  const classesPane = classNames({
    [`${classPrefix}-pane`]: true,
  })

  useEffect(() => {
    initData()
  }, [options])

  useEffect(() => {
    if (value !== state.innerValue) {
      state.innerValue = value as CascaderValue
    }
  }, [value])

  const initData = async () => {
    // 初始化开始处理数据
    state.lazyLoadMap.clear()
    if (format && Object.keys(format).length > 0) {
      state.optionsData = convertListToOptions(
        options as CascaderOption[],
        format as format
      )
    } else {
      state.optionsData = options
    }

    state.tree = new Tree(state.optionsData as CascaderOption[], {
      value: state.configs.optionKey.valueKey,
      text: state.configs.optionKey.textKey,
      children: state.configs.optionKey.childrenKey,
    })
    if (isLazy() && !state.tree.nodes.length) {
      await invokeLazyLoad({
        root: true,
        loading: true,
        text: '',
        value: '',
      })
    }
    state.panes = [
      {
        nodes: state.tree.nodes,
        selectedNode: null,
        paneKey: 'c1',
      },
    ]
    syncValue()
    setOptionsData(state.panes)
  }
  // 处理有默认值时的数据
  const syncValue = async () => {
    const currentValue = state.innerValue
    if (currentValue === undefined || !state.tree.nodes.length) {
      return
    }

    if (currentValue.length === 0) {
      state.tabsCursor = 0
      return
    }

    let needToSync = currentValue

    if (isLazy() && Array.isArray(currentValue) && currentValue.length) {
      needToSync = []
      const parent: any = state.tree.nodes.find(
        (node) => node.value === currentValue[0]
      )

      if (parent) {
        needToSync = [parent.value]
        state.initLoading = true

        const last = await currentValue
          .slice(1)
          .reduce(async (p: Promise<CascaderOption | void>, value) => {
            const parent = await p
            await invokeLazyLoad(parent)
            const node: any = parent?.children?.find(
              (item: any) => item.value === value
            )
            if (node) {
              needToSync.push(value)
            }
            return Promise.resolve(node)
          }, Promise.resolve(parent))
        await invokeLazyLoad(last)
        state.initLoading = false
      }
    }

    if (needToSync.length && currentValue === value) {
      const pathNodes = state.tree.getPathNodesByValue(needToSync)
      pathNodes.forEach((node, index) => {
        state.tabsCursor = index
        // 当有默认值时，不触发 chooseItem 里的 emit 事件
        chooseItem(node, true)
      })
    }
  }

  const invokeLazyLoad = async (node?: CascaderOption | void) => {
    if (!node) {
      return
    }

    if (!state.configs.onLoad) {
      node.leaf = true
      return
    }

    if (
      state.tree.isLeaf(node, isLazy()) ||
      state.tree.hasChildren(node, isLazy())
    ) {
      return
    }

    node.loading = true

    const parent = node.root ? null : node
    let lazyLoadPromise = state.lazyLoadMap.get(node)

    if (!lazyLoadPromise) {
      lazyLoadPromise = new Promise((resolve) => {
        // 外部必须resolve
        state.configs.onLoad?.(node, resolve)
      })
      state.lazyLoadMap.set(node, lazyLoadPromise)
    }

    const nodes: CascaderOption[] | void = await lazyLoadPromise

    if (Array.isArray(nodes) && nodes.length > 0) {
      state.tree.updateChildren(nodes, parent)
    } else {
      // 如果加载完成后没有提供子节点，作为叶子节点处理
      node.leaf = true
    }
    node.loading = false
    state.lazyLoadMap.delete(node)
  }

  const close = () => {
    onClose && onClose()
  }

  const closePopup = () => {
    close()
  }

  /* type: 是否是静默模式，是的话不触发事件
  tabsCursor: tab的索引 */
  const chooseItem = async (node: CascaderOption, type: boolean) => {
    if ((!type && node.disabled) || !state.panes[state.tabsCursor]) {
      return
    }
    // 如果没有子节点
    if (state.tree.isLeaf(node, isLazy())) {
      node.leaf = true
      state.panes[state.tabsCursor].selectedNode = node
      state.panes = state.panes.slice(0, (node.level as number) + 1)
      if (!type) {
        const pathNodes = state.panes.map((item) => item.selectedNode)
        const optionParams = pathNodes.map((item: any) => item.value)
        onChange(optionParams, pathNodes)
        onPathChange(optionParams, pathNodes)
      }
      setOptionsData(state.panes)
      close()
      return
    }
    // 如果有子节点，滑到下一个
    if (state.tree.hasChildren(node, isLazy())) {
      const level = (node.level as number) + 1

      state.panes[state.tabsCursor].selectedNode = node
      state.panes = state.panes.slice(0, level)
      state.tabsCursor = level
      state.panes.push({
        nodes: node.children || [],
        selectedNode: null,
        paneKey: `c${state.tabsCursor + 1}`,
      })
      setOptionsData(state.panes)
      setTabvalue(`c${state.tabsCursor + 1}`)

      if (!type) {
        const pathNodes = state.panes.map((item) => item.selectedNode)
        const optionParams = pathNodes.map((item: any) => item?.value)
        onPathChange(optionParams, pathNodes)
      }
      return
    }
    state.currentProcessNode = node
    if (node.loading) {
      return
    }

    await invokeLazyLoad(node)
    if (state.currentProcessNode === node) {
      state.panes[state.tabsCursor].selectedNode = node
      chooseItem(node, type)
    }
    setOptionsData(state.panes)
  }

  const renderItem = () => {
    return (
      <div className={`${classPrefix} ${className}`} style={style}>
        {popup && <div className="nut-cascader__title">{title}</div>}
        <Tabs
          value={tabvalue}
          title={() => {
            return optionsData.map((pane, index) => (
              <div
                onClick={() => {
                  setTabvalue(pane.paneKey)
                  state.tabsCursor = index
                }}
                className={`nut-tabs__titles-item ${
                  tabvalue === pane.paneKey
                    ? 'nut-tabs__titles-item--active'
                    : ''
                }`}
                key={pane.paneKey}
              >
                <span className="nut-tabs__titles-item__text">
                  {!state.initLoading &&
                    state.panes.length &&
                    pane?.selectedNode?.text}
                  {!state.initLoading &&
                    state.panes.length &&
                    !pane?.selectedNode?.text &&
                    '请选择'}
                  {!(!state.initLoading && state.panes.length) && 'Loading...'}
                </span>
                <span className="nut-tabs__titles-item__line" />
              </div>
            ))
          }}
        >
          {!state.initLoading && state.panes.length ? (
            optionsData.map((pane) => (
              <Tabs.TabPane key={pane.paneKey} value={pane.paneKey}>
                <div className={classesPane}>
                  {pane.nodes &&
                    pane.nodes.map((node: any, index: number) => (
                      <CascaderItem
                        key={index}
                        {...props}
                        data={node}
                        checked={pane.selectedNode?.value === node.value}
                        chooseItem={(node: any) => chooseItem(node, false)}
                      />
                    ))}
                </div>
              </Tabs.TabPane>
            ))
          ) : (
            <Tabs.TabPane>
              <div className={classesPane} />
            </Tabs.TabPane>
          )}
        </Tabs>
      </div>
    )
  }

  return (
    <>
      {popup ? (
        <Popup
          // className="nut-cascadar-popup"
          visible={visible}
          position="bottom"
          round
          closeIcon={closeIcon}
          closeable={closeable}
          closeIconPosition={closeIconPosition}
          // todo 只关闭，不处理逻辑。和popup的逻辑不一致。关闭时需要增加是否要处理回调
          onClickOverlay={closePopup}
          onClickCloseIcon={closePopup}
        >
          {renderItem()}
        </Popup>
      ) : (
        renderItem()
      )}
    </>
  )
}

export const Cascader = React.forwardRef(InternalCascader)

Cascader.defaultProps = defaultProps
Cascader.displayName = 'NutCascader'
