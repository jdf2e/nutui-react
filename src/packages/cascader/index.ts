import { Cascader } from './cascader'

export type {
  CascaderPane,
  CascaderFormat,
  CascaderOptionKey,
  CascaderOption,
  CascaderValue,
  CascaderConfig,
} from './types'
export type { CascaderProps } from './cascader'
export default Cascader
