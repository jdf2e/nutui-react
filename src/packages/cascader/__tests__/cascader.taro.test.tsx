import TestUtils from '@tarojs/test-utils-react'
import '@testing-library/jest-dom'
import { Cascader } from '../cascader'

const testUtils = new TestUtils()
const mockOptions = [
  {
    value: '浙江',
    text: '浙江',
    children: [
      {
        value: '杭州',
        text: '杭州',
        disabled: true,
        children: [
          { value: '西湖区', text: '西湖区' },
          { value: '余杭区', text: '余杭区' },
        ],
      },
      {
        value: '温州',
        text: '温州',
        children: [
          { value: '鹿城区', text: '鹿城区' },
          { value: '瓯海区', text: '瓯海区' },
        ],
      },
    ],
  },
  {
    value: '湖南',
    text: '湖南',
    disabled: true,
  },
  {
    value: '福建',
    text: '福建',
    children: [
      {
        value: '福州',
        text: '福州',
        children: [
          { value: '鼓楼区', text: '鼓楼区' },
          { value: '台江区', text: '台江区' },
        ],
      },
    ],
  },
]
const mockKeyConfigOptions = [
  {
    name: '浙江',
    items: [
      {
        name: '杭州',
        disabled: true,
        items: [{ name: '西湖区' }, { name: '余杭区' }],
      },
      {
        name: '温州',
        items: [{ name: '鹿城区' }, { name: '瓯海区' }],
      },
    ],
  },
  {
    name: '湖南',
    disabled: true,
  },
  {
    name: '福建',
    items: [
      {
        name: '福州',
        items: [{ name: '鼓楼区' }, { name: '台江区' }],
      },
    ],
  },
]
const mockConvertOptions = [
  { value: '北京', text: '北京', nodeId: 1, nodePid: 0, sort: 2 },
  { value: '朝阳区', text: '朝阳区', nodeId: 11, nodePid: 1 },
  { value: '亦庄', text: '亦庄', nodeId: 111, nodePid: 11 },
  { value: '广东省', text: '广东省', nodeId: 2, nodePid: 0, sort: 1 },
  { value: '广州市', text: '广州市', nodeId: 21, nodePid: 2 },
]
const execFn = (desc: string, component: any, props: object, cb: any) => {
  it(desc, async () => {
    await testUtils.mount(component, {
      props,
    })
    cb()
    testUtils.unmout()
  })
}
describe('cascader-taro', () => {
  execFn(
    'options',
    Cascader,
    {
      value: ['福建', '福州', '鼓楼区'],
      options: mockOptions,
    },
    async () => {
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
  execFn(
    'options with valueKey/textKey/childrenKey',
    Cascader,
    {
      value: ['福建', '福州', '鼓楼区'],
      options: mockKeyConfigOptions,
      optionKey: {
        valueKey: 'name',
        textKey: 'name',
        childrenKey: 'items',
      },
    },
    async () => {
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
  execFn(
    'options with format',
    Cascader,
    {
      value: ['广东省', '广州市'],
      options: mockKeyConfigOptions,
      format: {
        topId: 0,
        idKey: 'nodeId',
        pidKey: 'nodePid',
        sortKey: 'sort',
      },
    },
    async () => {
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
  execFn(
    'visible false',
    Cascader,
    {
      value: ['广东省', '广州市'],
      visible: false,
      options: mockOptions,
    },
    async () => {
      expect(testUtils.queries.querySelector('.nut-popup')).toBeNull()
    }
  )
  execFn(
    'visible true',
    Cascader,
    {
      value: ['广东省', '广州市'],
      visible: false,
      options: mockOptions,
    },
    async () => {
      expect(testUtils.queries.querySelector('.nut-popup')).toBe
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
  execFn(
    'value',
    Cascader,
    {
      options: mockOptions,
    },
    async () => {
      expect(
        testUtils.queries.querySelectorAll(
          '.nut-cascader-item[aria-checked="true"]'
        ).length
      ).toBe(0)
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
  execFn(
    'init Value without defaultValue',
    Cascader,
    {
      visible: true,
      value: ['福建', '福州', '鼓楼区'],
      options: mockOptions,
    },
    async () => {
      const element = testUtils.queries.querySelectorAll(
        '.active.nut-tabpane .active .nut-cascader-item-title'
      )[0]
      expect(element).toHaveTextContent('鼓楼区')
    }
  )
  execFn(
    'init Value with both value and defaultValue',
    Cascader,
    {
      visible: true,
      value: ['福建', '福州', '台江区'],
      defaultValue: ['福建', '福州', '鼓楼区'],
      options: mockOptions,
    },
    async () => {
      const element = testUtils.queries.querySelectorAll(
        '.active.nut-tabpane .active .nut-cascader-item-title'
      )[0]
      expect(element).toHaveTextContent('台江区')
    }
  )
})
