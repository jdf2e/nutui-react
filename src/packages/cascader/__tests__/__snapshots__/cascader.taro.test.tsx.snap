// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cascader-taro options 1`] = `
"[36m<body>[39m
  [36m<div />[39m
[36m</body>[39m"
`;

exports[`cascader-taro options with format 1`] = `
"[36m<body>[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
[36m</body>[39m"
`;

exports[`cascader-taro options with valueKey/textKey/childrenKey 1`] = `
"[36m<body>[39m
  [36m<div />[39m
  [36m<div />[39m
[36m</body>[39m"
`;

exports[`cascader-taro value 1`] = `
"[36m<body>[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
[36m</body>[39m"
`;

exports[`cascader-taro visible true 1`] = `
"[36m<body>[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div />[39m
[36m</body>[39m"
`;

exports[`options 1`] = `
"[36m<body>[39m
  [36m<div />[39m
[36m</body>[39m"
`;
